src/simple_rtt.rs

```diff
@@ -0,0 +1,201 @@
+    /// request.
+    pub fn default_port(mut self, port: u16) -> Self {
+        self.tripper.default_port = port;
+        self
+    }
+
+    /// Sets the timeout after which requests will abort if they aren't finished
+    pub fn timeout(mut self, timeout: Duration) -> Self {
+        self.tripper.timeout = timeout;
+        self
+    }
+
+    /// Builds the final `Tripper`
+    pub fn build(self) -> Tripper {
+        self.tripper
+    }
+}
+
+impl Tripper {
+    /// Construct a new `Tripper` with default parameters
+    pub fn new() -> Self {
+        Tripper::default()
+    }
+
+    /// Returns a builder for `Tripper`
+    pub fn builder() -> Builder {
+        Builder::new()
+    }
+}
+
+/// Try to read a line from a buffered reader. If no line can be read till the deadline is reached
+/// return a timeout error.
+fn get_line<R: BufRead>(reader: &mut R, deadline: Instant) -> Result<String, Error> {
+    let mut line = String::new();
+    while deadline > Instant::now() {
+        match reader.read_line(&mut line) {
+            // EOF reached for now, try again later
+            Ok(0) => std::thread::sleep(Duration::from_millis(5)),
+            // received useful data, return it
+            Ok(_) => return Ok(line),
+            // io error occurred, abort
+            Err(e) => return Err(Error::SocketError(e)),
+        }
+    }
+    Err(Error::Timeout)
+}
+
+impl HttpRoundTripper for Tripper {
+    type ResponseBody = Cursor<Vec<u8>>;
+    type Err = Error;
+
+    fn post(&self, request: http::Request<&[u8]>) -> Result<http::Response<Self::ResponseBody>, Self::Err> {
+        // Parse request
+        let str_port = (
+            match request.uri().host() {
+                Some(s) => s,
+                None => return Err(Error::NoHost),
+            },
+            request.uri().port_u16().unwrap_or(self.default_port),
+        );
+        let server = str_port.to_socket_addrs()?.next().unwrap();
+
+        let method = request.method();
+        let uri = request.uri().path_and_query().map(|p| p.as_str()).unwrap_or("/");
+
+        // Open connection
+        let request_deadline = Instant::now() + self.timeout;
+        let mut sock = TcpStream::connect_timeout(&server, self.timeout)?;
+
+        // Send HTTP request
+        sock.write_all(format!("{} {} HTTP/1.0\r\n", method, uri).as_bytes())?;
+        for (key, value) in request.headers() {
+            sock.write_all(key.as_ref())?;
+            sock.write_all(b": ")?;
+            sock.write_all(value.as_ref())?;
+            sock.write_all(b"\r\n")?;
+        }
+        sock.write_all(b"\r\n")?;
+        sock.write_all(request.body())?;
+
+        // Receive response
+        let mut reader = BufReader::new(sock);
+
+        // Parse first HTTP response header line
+        let http_response = get_line(&mut reader, request_deadline)?;
+        if http_response.len() < 12 || !http_response.starts_with("HTTP/1.0 ") {
```

Comment:
Wouldn't `HTTP/1.1` be a better choice? I would imagine that's the most commonly used version, and some servers might just hard-code 1.1 response, instead of downgrading to 1.0 because they don't expect anyone to call with 1.0. Requesting 1.1 and accepting both 1.1 and 1.0 seems like not a lot of complexity with some additional compatibility.

Reply:
Sure, added a commit (and checked that the Liquid integration tests pass when using this to do jsonrpc)